src/shell/Makefile.am

```diff
@@ -87,7 +87,8 @@ flux_shell_SOURCES = \
 	mustache.c \
 	doom.c \
 	exception.c \
-	rlimit.c
+	rlimit.c \
+	unshare.c
```

Comment:
Well... I'm not sure if this actually fixes anything.   This shell plugin doesn't do anything that a wrapper program like unshare(1) or other more sophisticated container tooling can't do.  It relies on the system being configured to allow user namespaces.  As @trws said in #3927
>  If user namespaces for regular users are enabled by default on TOSS4, then we probably don't need to do anything in flux at all.

So this is a solution in search of a problem.  

I found it interesting to play around with.  If it were feasible to run jobs with `-o unshare=user,maproot,mount` which gives the shell CAP_SYS_ADMIN capability, then the shell could do stuff like set up a private `/flux` mount for the "jobtmp" or maybe eventually for a synthetic proc-like file system with directories for running jobs.  Or maybe for I/O fowrarding from rank 0 :-).

I can't quite tell if this goes beyond my personal learning tool so I was hesitant to put it forward without any kind of plan for using it to do something useful.  OTOH it's not a lot, and it's completely standalone.

Reply:
As it turns out, user namespaces are *not* enabled by default on TOSS4, we actually still need something that sets up uid mappings to make that all work.  I believe Ryan mentioned someone was looking at porting the slurm plugin for this, but that is not done yet AFAIK.  The main thing this would give us, is an easy way to do user mounts of things like tempfs, bind mounts, etc.  We could also use it to create a process namespace to ensure we could reap all the child processes more easily.

Reply:
Tom and I had an offline discussion about this, and it appears that our TOSS4 Flux systems at least enable this:
```
# sysctl user.max_user_namespaces
user.max_user_namespaces = 16384
```
which is sufficient for this plugin to do what it claims including mapping root to the job user and within the job, mounting tmpfs and binding filesystems.  In theory the shell should be able to do all the things for a job that the unprivileged [unshare(1)](https://man7.org/linux/man-pages/man1/unshare.1.html) can do, and it is able create a process namespace.  ( This is tricky to set up, involving synchronization across fork(), so this PR doesn't try to do that in the shell, but FWIW it should be possible without additional privilege.)

I'm not sure what doesn't work but we could use unshare(1) to probe that.  @trws mentioned the ability to map multiple user IDs back to the job user is needed so that a full OS container could be started up.  I haven't figured out how to try that in unshare(1) yet.

Reply:
One open question for me is how a separate `unshare` plugin will work with "downstream" plugins that require or assume some form of `unshare(2)` has already been used. How will the dependent plugins determine that the user activated the `unshare` plugin with the required arguments? Does it make more sense, or is it even possible, for plugins that require namespaces to call `unshare(2)` themselves, so that more meaningful options can be provided on the command line? I guess the main point is that it is difficult to determine if this plugin has the right design or is the right fit without a use case available.

That being said, I understand that this plugin may allow easier experimentation via a shell option rather than just wrapping a command in `unshare(1)`. The only drawback here is that once the plugin is added a shell builtin, we have to maintain it, deal with feature requests, etc. Again, this is a slightly harder sell when there is no active use case for the plugin.

I'm not strongly arguing against inclusion of the plugin, I just wanted to bring up the points above.

Reply:
Yeah agreed.  I'm fine with closing this PR for now.  It seems like if we want to enable stuff like shell plugins mounting private namespaces, we'll need coordination between plugins that are requesting that different namespaces be unshared.  We may also need some support in libsubprocess for inserting some proc poking and synchronization before and after fork/clone/posix_spawn.